scheduler/remote_server.py

```diff
@@ -0,0 +1,487 @@
+                                    queued_lines.append(item_to_add)
+                                else:
+                                    queued_lines.append(item_to_add)
+                            else:
+                                # if the prio is lower, then we only the schedule the piece that
+                                # doesn't overlap.
+                                item_to_add['time'] = queued_finish
+                                new_duration = finish_time - queued_finish
+                                item_to_add['duration'] = new_duration.total_seconds()//60
+                                queued_lines.append(item_to_add)
+                    else:
+                        time_diff = item_to_add['time'] - queued_finish
+                        new_line = copy.deepcopy(inf_dur_line)
+                        new_line['duration'] = time_diff.total_seconds()//60
+                        new_line['time'] = queued_finish
+                        queued_lines.append(new_line)
+                        queued_lines.append(item_to_add)
+
+        if idx == len(scd_lines) - 1:
+            last_tt_line, queued_finish = calculate_new_last_line_params()
+
+            inf_dur_line['time'] = queued_finish
+            queued_lines.append(inf_dur_line)
+
+    queued_dict = collections.OrderedDict()
+    for line in queued_lines:
+        if not line['order'] in queued_dict:
+            queued_dict[line['order']] = []
+
+        queued_dict[line['order']].append(line)
+
+    return queued_dict
+
+def timeline_to_atq(timeline, scd_dir, now):
+    """ Converts the created timeline to actual atq commands.
+
+    Args:
+        timeline (dict): A dictionary holding all timeline events.
+        scd_dir (str): The directory with SCD files.
+        now (datetime): The datetime holding the time of scheduling.
+
+    Log and backup the existing atq, remove old events and then schedule everything recent. The
+    first entry should be the currently running event, so it gets scheduled immediately.
+    """
+    backup_cmd = 'for j in $(atq | sort -k6,6 -k3,3M -k4,4 -k5,5 |cut -f 1);'\
+    'do atq |grep -P "^$j\t"; at -c "$j" | tail -n 2; done'
+
+    output = sp.check_output(backup_cmd, shell=True)
+
+    backup_time_str = now.strftime("%Y.%m.%d.%H.%M")
+    backup_dir = "{}/atq_backups".format(scd_dir)
+
+    if not os.path.exists(backup_dir):
+        os.makedirs(backup_dir)
+
+    backup_file = "{}/{}.atq".format(backup_dir, backup_time_str)
+
+    with open(backup_file, 'wb') as f:
+        f.write(output)
+
+    clear_command = "for i in `atq | awk '{print $1}'`;do atrm $i;done"
+
+    sp.call(clear_command, shell=True)
+
+    atq = []
+    first_event = True
+    for _,events in timeline.items():
+        for event in events:
+            if first_event:
+                atq.append(format_to_atq(event['time'], event['experiment'], True))
```

Comment:
This is true, but I think this adds a bunch of complexity to this stage. Cause we would have to look at the atq, or a backup of the SCD to find the currently running experiment and recalculate durations to see if the experiment matches. This is making me think though that any manual changes to the scd should be done in a staging area before being moved to the dir that syncs to the site. We dont want to trigger a change in scheduling for every single line that gets added/removed during scheduling of discretionary time.

Reply:
My feeling is that we should resolve this later (create a new issue for improvement of this). Isn't the last line in the schedule always an infinite duration line (making the duration calc easier)? So then it would just have be entered as the first infinite duration line in convert_scd_to_timeline. Either way, can be improved later.

Reply:
I agree, I think the complications occur if that line is removed or modified.